### DISCUSSION TITLE
Visual editor for elliptical arcs

### DISCUSSION THREAD

**Discussion by Dimon-II** [Yqnn/svg-path-editor#79] [Ideas]
![image](https://user-images.githubusercontent.com/29166758/167499137-bc4d02aa-0d05-4b1e-bad4-4290240ff918.png)
I have been looking for a simple graphic editor for SVG path for a long time. This is exactly what is missing in my program Decard. 
I really liked the idea of ​​quickly switching relative and absolute coordinates, this is something that I myself did not guess.
I'm also at a loss how to make a visual editor for elliptical arcs. When selecting a segment, you can display the entire ellipse, but how to change the radii and the angle of inclination... convenient and simple... 
Enclose ellipse in a rectangle? 
Show all 4 possible arcs with mouse selection?

By the way, when I inserted an elliptical SVG path segment into a font editor, it was automatically converted to simple C-arcs.

**Comment by Yqnn** [comment]
I did several tests to improve the visual edition of elliptical arcs, but I didn't get any satisfying result yet.
I think drawing the full ellipsis with semi-major and semi-minor axes is a must.

> By the way, when I inserted an elliptical SVG path segment into a font editor, it was automatically converted to simple C-arcs.

Ellipsis can only be approximated using Bezier curves, but I guess font formats don't support ellipsis because it's too CPU intensive.

**Comment by Yqnn** [comment]
https://user-images.githubusercontent.com/588624/168797811-c234fb53-4503-4e7f-8131-0f7f6f0fc399.mp4

**Reply by Dimon-II** [reply]
Is only angle setup. The problem is that these radii cannot be changed with the mouse - when you change any, the center also shifts!

**Reply by Yqnn** [reply]
I think it's possible to change minor and major axis by dragging the center of the ellipse (that could also swap the large-arc flag automatically).
But that would involve heavy equations 😅

**Reply by Yqnn** [reply]
My preliminary test is convincing :)

https://user-images.githubusercontent.com/588624/168858785-a755d916-6a85-40a3-9d6f-8bc3b7b67202.mp4

**Reply by Dimon-II** [reply]
All you have to do is switch the arcs.  Clickable dots at the ends of diameter lines?

And I have to think how to repeat these manipulations of yours with Delphi programming...

**Comment by Yqnn** [comment]
Letting the center of the ellipsis freely movable is actually a bad option: depending on the points configuration it may behave very poorly.

I will try something else: allowing to move the ellipsis center in a way that preserves the rx/ry ratio.
Then I will put 2 other control points for rx and ry.

**Comment by Yqnn** [comment]
I finally have something that is usable in all situations:


https://user-images.githubusercontent.com/588624/169618692-25630b6f-3578-4565-8c08-8a7d2ab3cf14.mp4

**Reply by Dimon-II** [reply]
The non-standard interface is that when changing the radius, the mouse cursor loses contact with the controlled point that marks the radius. 

Maybe show the radius vectors not from the center of the ellipse, but from the start or end point of the path?

**Reply by Yqnn** [reply]
The trajectory of radius points is pretty unpredictable, there is the same issue with the angle.
I think the best option would be to display some visual hints on how to move the cursor.

Examples of weird trajectories:

https://user-images.githubusercontent.com/588624/169668962-fdcfd602-9b90-498d-8927-bbf03f5ba64d.mp4

https://user-images.githubusercontent.com/588624/169668967-3152c68d-6492-477d-a37a-3e14f867c04b.mp4
